src/hotspot/share/classfile/stringTable.hpp

```diff
@@ -56,6 +56,38 @@ class StringTable : AllStatic {
   static double get_load_factor();
   static double get_dead_factor(size_t num_dead);
 
+  enum class StringType {
+    OopStr, UnicodeStr, SymbolStr, UTF8Str
+  };
+
+  struct StringWrapperInternal {
+    union {
+      const Handle oop_str;
+      const jchar* unicode_str;
+      const Symbol* symbol_str;
+      const char* utf8_str;
+    };
+    const StringType type;
+    const size_t length;
+
+    StringWrapperInternal(const Handle oop_str, const size_t length)     : oop_str(oop_str),         type(StringType::OopStr), length(length)     {}
+    StringWrapperInternal(const jchar* unicode_str, const size_t length) : unicode_str(unicode_str), type(StringType::UnicodeStr), length(length) {}
+    StringWrapperInternal(const Symbol* symbol_str, const size_t length) : symbol_str(symbol_str),   type(StringType::SymbolStr), length(length)  {}
+    StringWrapperInternal(const char* utf8_str, const size_t length)     : utf8_str(utf8_str),       type(StringType::UTF8Str), length(length)    {}
+  };
+
+public:
+  typedef struct StringWrapperInternal StringWrapper;
+
+  // Unnamed int needed to fit CompactHashtable's equals type signature
+  static bool wrapped_string_equals(oop java_string, StringWrapper wrapped_str, int = 0);
+
+private:
+  static const char* get_symbol_utf8(StringWrapper symbol_str);
+  static unsigned int hash_wrapped_string(StringWrapper wrapped_str);
+  static const jchar* to_unicode(StringWrapper wrapped_str, int &len, TRAPS);
+  static Handle handle_from_wrapped_string(StringWrapper wrapped_str, TRAPS);
```

Comment:
Why is StringWrapper passed by value not by reference? If it's passed by reference, could you hide this whole class in stringTable.cpp ?

Reply:
No reason in particular. I changed the functions to use references instead and hid the wrapper away in the .cpp file.